src/platform_impl/android/mod.rs

```diff
@@ -144,8 +144,24 @@ impl<T: 'static> EventLoop<T> {
                             event::Event::Suspended
                         );
                     }
-                    Event::Pause => self.running = false,
-                    Event::Resume => self.running = true,
+                    Event::Pause => {
+                        call_event_handler!(
+                            event_handler,
+                            self.window_target(),
+                            control_flow,
+                            event::Event::Suspended
+                        );
+                        self.running = false;
+                    }
+                    Event::Resume => {
+                        call_event_handler!(
+                            event_handler,
+                            self.window_target(),
+                            control_flow,
+                            event::Event::Resumed
+                        );
+                        self.running = true;
+                    }
```

Comment:
Can you please add some information about why this is necessary, maybe with a few links to the Android docs? `WindowDestroyed` already sends `Suspended`, why is that not enough?

Reply:
https://github.com/rust-windowing/glutin/issues/904
https://github.com/rust-windowing/glutin/pull/1313

Check this comment:
https://github.com/rust-windowing/glutin/pull/1313#issuecomment-656477993

Android has some special windows context handling, where context cannot be resused after resuming. Data has to be reinitialized. Therefore we need to handle those events.

Old version just sets flags:
```
Event::Pause => self.running = false,
Event::Resume => self.running = true
```

Reply:
More discussion on this:
https://github.com/rust-windowing/android-ndk-rs/issues/117

Reply:
I doubt that is correct: https://developer.android.com/guide/components/activities/activity-lifecycle#onpause

Reply:
> I doubt that is correct: https://developer.android.com/guide/components/activities/activity-lifecycle#onpause
 Why? We need to handle pause and resume events on android side. The point is, no one can tell how long the app was paused, maybe it was send to background for several hours... We need to free graphics resources here. If the app resumes, it's possible that original window context is not valid anymore. We cannot keep rendering to that destroyed window.
 
 Just settings flags is not an option.

Reply:
The linked docs are pretty clear on this:

> You can also use the onPause() method to release system resources, handles to sensors (like GPS), or any resources that may affect battery life while your activity is paused and the user does not need them. However, as mentioned above in the onResume() section, a Paused activity may still be fully visible if in multi-window mode. As such, you should consider using onStop() instead of onPause() to fully release or adjust UI-related resources and operations to better support multi-window mode.

Reply:
In other words: while I do agree it might be nice to forward these events through `winit`, the `Suspended` and `Resumed` events are not right as they serve a different purpose.

`onPause` and `onResume` should result in a [`WindowEvent::Focused`](https://docs.rs/winit/latest/winit/event/enum.WindowEvent.html#variant.Focused).

Perhaps we should rename `Suspended` and `Resumed` to more closely represent the graphics state needing to be forcibly destroyed and recreated, respectively?

Reply:
I agree, maybe I abused these events to get my context handling correctly. The main point is, that we must not lose important android events. We need to be able to react on window changes.
When do resourced need to be created, when do they need to be destroyed.

And yes you are right, it would be much more coherent, if these events would be of type `WindowEvent`. Like i'm handling the resize event:

```
Event::WindowEvent { event, .. } => match event {
    WindowEvent::Resized(physical_size) => {
        if let Some(device_ctx) = self.device_ctx.as_mut() {
            device_ctx.window_context.resize(physical_size);

            // call resize device callback
            self.runner.resize_device(
                &device_ctx.gl,
                physical_size.width,
                physical_size.height,
            );
        }
    }
    ....
```

Reply:
I checked the android behaviour:
Focus can not be used to create the device, as i get following events when starting the main loop:

- resume
- resize
- focus

There is no window create event or such. The focus event is too late to use it for creation, as resize event already executed.

Reply:
> Focus can not be used to create the device

Nor is the android documentation suggesting to. Swapchain creation and deletion "must" be tied to window creation and destroying: you can do it whenever you like, as long as it is created _after_ `WindowCreated` and destroyed before or during `WindowDestroyed`. Applications are free to perform this elsewhere, as long as they synchronize properly with the events (ie. hold on to the lock from `native_window()`, and destroy your resources followed by releasing this lock whenever `WindowDestroyed` is called).

> There is no window create event or such.

At what level are you logging this? Is logging initialized early enough (ie. in `ndk_glue::main`)? If Android doesn't give your application a window there's nothing to render to.
My best guess is "resume" in your list is referring to `event::Event::Resumed`, ie. triggered on `ndk_glue::Event::WindowCreated`?